### DISCUSSION TITLE
a 'by' keyword for interface auto implementation

### DISCUSSION THREAD

**Discussion by fat-rooster** [dotnet/csharplang#7026] [Language Ideas]
For many interfaces there are relationships between the functions that should be common between implementations (or at least are between many). Often certain functions can be expressed as compositions of other functions in the interface. In these cases we could write an implementation as some subset* of the implementation, together with a source generator to extend that to the rest of the interface.
A toy example for clarity:
`interfaceIMult<T> where T: IMult<T>
{
T Mult(T b);
T Pow(int b);
}`
If MultToPow is a source generator that creates a power function from any multiplication then we could write:
`class dumbMult : IMult<dumbMult> by MultToPow`
And it would automatically generate our integer power function from a supplied Mult function.

In this particular case the logic for minimising the number of multiplications can produce impressive performance gains, so we actively do not want implementers rolling their own.
In addition, being able to supply a default* auto-implementation function that operates if the interface is not fully implemented (after a supplied extender is run) could partially solve backwards compatibility issues, but also make the path of least resistance for programmers be an intended auto-implementation (in our toy this might be appropriate).

* 'default' does not have to mean designed by the interface designer. It could be overwritten. The point is that the interface designer **can** provide a default, making using it just a case of grabbing it 'off the shelf' and placing it.

Of note is that a generator could repeatedly call itself on a class with any subset of the interface, skipping functions that it doesn't have enough to build yet. It either detects that it cannot build any more functions, and fails, or continues building until it has a complete implementation. It would be able to build a full implementation from a whole variety of different 'basis' sets of implemented functions. The interface designer does not need to know what the natural representation of the implementation is.

In terms of details; an actual source generator is probably not how you would want to expose this feature. For that I would suggest an attribute on an abstract class. It wants to be abstract because it should never actually exist. All it's functions are going to be circularly dependent, if they are defined at all. One issue with that is that we may want multiple definitions of a function for it to choose from, so creating a more dedicated structure might be warranted.

In terms of issues, the obvious one is hiding bugs. If a user intends to explicitly implement something, but makes a typo, the generator may carry on and just hide the error. It may not even be caught as an unused function, if the user references it in their own code. There are probably more

This connects with my other suggestion on interfaces including tests in that it can let the interface designer provide 'canonical' implementations that describe how the various functions should relate, which could form the basis of auto generated test frameworks for implementations.

I'm pretty new to this, so may be missing something obvious.

**Comment by En3Tho** [comment]
Why do you need a new special keyword if you can just splat an attribute on a class to have the same result? (Meaning source generator find an attribute and provides interface implementation)

**Reply by fat-rooster** [reply]
An attribute doesn't relate to the interface it affects in any visual way. The code would be 
```
[the attributes that describe how the interfaces are implemented]
class name: all the interfaces.
```
It would be much cleaner if each interface implementation could be put beside the interface name.

```
class name : interface by x, interface by y...
```

but yes, the functionality is absolutely possible already. I'm suggesting it should be more nomalised by being supported by dedicated syntax.

**Reply by En3Tho** [reply]
But you can clearly describe it via generic attributes in the same way. I dont feel like substituting one syntax for another makes much sense

**Reply by fat-rooster** [reply]
See the thread below for a longer example that explains my position better, but I'm under no illusions that it is a world shattering improvement. It should be pretty simple to do though, and I don't see what else it could conflict with syntactically.

**Reply by En3Tho** [reply]
Well, I saw it and I can just repeat myself: I don't feel like swapping one syntax for another makes much sense:
1. Very similar amount of keystrokes as attribute approach
2. Doesn't really help gracefully shortening the code to do a giant thing into just a few small things, you know
3. What you want is a small subset of what source generators can already do

Now the language cost:
1. New keyword for the very narrow scope of what is being provided by it
2. It's in the language forever, no new feature can use it even if it is better than current one
 
To summarize, it feels like too great cost for too little profit. To introduce a new keyword to the language you have to prove that this feature you want really needs it and feature itself is very compelling to the language and clearly improves certain aspects of it that are "weak" now

**Reply by fat-rooster** [reply]
'by' is already a contextual keyword. Linq uses it. I'm just suggesting another context for it; an extremely narrow one that is never used for anything else. Part of the reason for the contextual keyword between the interface and the implementation is specifically to avoid blocking the context. If someone wanted to propose that the code
` class C: I in potato`
My suggestion would not block it. I have no idea what such code would even mean, but I also cannot think what else the 'by' keyword could reasonably mean in that context. There are some open questions on the best way to resolve conflicts, but the feature itself is very simple. Simply mapping it to attributes would work fine.

'We can already do that with source generators' isn't an answer to the question 'is this a better way of doing this?'. All our high level features are simply sugar for things we can already do. Granted, I don't think it is much better, but I think you are overestimating the cost.

**Comment by HaloFour** [comment]
You don't provide any examples of how this `by` keyword would be used.

The common implementations, especially as composition between other members, can already be achieved through default implementations.

Otherwise, the rest of this sounds a lot like interface delegation to me: #234

**Reply by fat-rooster** [reply]
Interface delegation would be a special case of this, but this can do much more. It is also more general than default implementations, because you can still specify the implementations, just whole swathes of them at a time.
As a specific example of how it might look:
```
interfaceIMult<T> where T: IMult<T> 
{
  T Mult(T a);
  T Pow(int power); 
}
class DumbMult : IMult<DumbMult> by MultToPow
{
  public float X;
  public DumbMult(x){X=x;)
  public DumbMult Mult(DumbMult a)   
  {
    return new DumbMult(X*a.X);
  }
}
```
Together with that, you would also need some way of building MultToPow. People comfortable writing source generators do not need this sort of feature anyway, so we would want something simple, yet familiar. For example:
```
[InterfaceImplementation(typeof(IMult))]
abstract class MultToPow : IMult<MultToPow >
{
  public abstract MultToPow Mult(MultToPow  b);
  public MultToPow Pow(int power)
  {
   var rv=Mult(this);
   for(int i=2; i<power; i++)
   {
      rv.Mult(this)
   }
   return rv;
  }
}
```
The attribute would be responsible for building the source generator. This method of describing the source generator has some down sides, so I'm not super attached to it, but some easy way of building the source generators would be a big part of making the feature user friendly.

As for how it relates to interface delegation, you could easily just write a delegatesTo source generator, and then it would become
```
class Parent : IInterface by delegatesTo("Child")
{
  IInterface Child;
  public void SomeInterfaceFunctionToCallFromHere(){}
}
```

Is that clearer?

**Reply by HaloFour** [reply]
Why does this need a new feature?  You can already declare a new interface that extends the first interface and provides default implementations of its members.

**Reply by fat-rooster** [reply]
Because then you need to define a different interface for every possible basis of the interface, and users need to know what the name of the particular interface that has the defaults they want. When people are reading your code they might see an interface that they don't think they know, but is actually an alias for one that they do. That seems much less clear.

The biggest difference from defaults is that this is pure compile time, and that makes it pretty easy. It is about clean code and avoiding boilerplate, not backwards compatibility. As pointed out above, it is already possible with attributes, so it can easily be lowered to an attribute. The way we build the source generators is up for debate (though permitting multiple answers is part of the point), but the actual 'by' keyword is an extremely low cost feature that I think could make for very clean code.

The other major difference from defaults is that it is possible to use non interface members as the basis functions too. Converting from another interface that does something similar could be automated using the same tool chain.

**Reply by CyrusNajmabadi** [reply]
The attribute form already seems like very clean code, and is already available today.  It would definitely be good to know why that approach is not sufficient.  Thanks! :)

**Reply by HaloFour** [reply]
@fat-rooster 

> Because then you need to define a different interface for every possible basis of the interface, and users need to know what the name of the particular interface that has the defaults they want. When people are reading your code they might see an interface that they don't think they know, but is actually an alias for one that they do. That seems much less clear.

I don't see why it's any less clear than having the definitions in completely disparate types.  You still need to know the name, and you need to know how to compose them.  Except that with a source generator in the middle the relationship is a lot less clear.

> The biggest difference from defaults is that this is pure compile time, and that makes it pretty easy.

Why is that a virtue?  Default members are already a part of the runtime, which makes it even easier than a new feature.

> It is about clean code and avoiding boilerplate, not backwards compatibility.

Default implementations are not just for backwards compatibility.  They are a perfectly useful mechanism for implementing traits, and composition of traits.

Lastly, if the goal is to enable a source generator to automatically implement the members, there's zero need for a language feature here.  You can already use attributes to declare that relationship.  I would argue that syntax should _not_ be added unless it is in support of first-class features within the language.

**Reply by fat-rooster** [reply]
@CyrusNajmabadi 
Attributes are pretty clean when dealing with a couple of interfaces, but it gets slightly unwieldy if there are many. Compare:
```
[UnrelatedAttribute1]
[UnrelatedAttribute2]
[I1extension]
[extensionOfI2]
[I4extension]
[DelegatesTo<I5>("Child")]
class C : I1, I2, I3, I4, I5 
{...I2And5 Child; ...}
```
with
```
[UnrelatedAttribute1]
[UnrelatedAttribute2]
class C:
    I1 by I1extension,
    I2 by extensionOfI2,
    I3, 
    I4 by I4extension,
    I5 by DelegatesTo("Child")
{...I2And5 Child; ...}
```
I would argue that the second is cleaner, though the first is still pretty good. In particular, if you have generic methods for implementing an interface you will need to repeat the type in the attribute anyway. See I5 in the above example. The child may implement many of the interfaces, but we don't want to delegate to it for them. Additionally, the only thing letting you know how each interface is being implemented in the first example is because I named them to make that clear. In the second it is more explicit (though funny business with overlapping functions may still occur).

The attribute approach is definitely sufficient, but that doesn't mean we can't do better with a little syntactic sugar. To my knowledge there is nothing that can come between an interface and the comma, so this can't break anything. I can't imagine anything else that might want to go there either.

**Reply by HaloFour** [reply]
I'd much rather have a conversation about exposing more attribute targets, either in the runtime or via [source-only attributes](https://github.com/dotnet/csharplang/discussions/1452).  I don't think a new syntax for custom metadata to be used only to drive source generators belongs in the language at all, especially using new keywords.  That should be reserved for proper language features.

**Reply by fat-rooster** [reply]
@HaloFour the reason we would not want to inherit is that people downstream would see a interface that is inherited purely for ease of implementation. It is exposing an implementation detail which is what interfaces are designed to hide. It is not a new interface, it should not have a new name.
As for knowing the name, in my version the writer needs to know the name, but a reader does not. If you were to extend the interface instead, by convention it might be named something like 'IInterfaceXByWhatever', and reasonably clear it inherits from IInterfaceX, but nothing enforces that. Somebody looking at the class who just wants to know what it does is, and so is looking at the interfaces it implements is hindered by that pattern. In mine they see the interface they know, even if they are unfamiliar with the generator that implements most of it.
You could argue that my suggestion is also polluting the class header with implementation details, but at least it does so without obfuscating the actual interfaces being implemented.

I am well aware that there is no 'need' for this feature, any more than there was any need for any feature that removes boilerplate. It would be just a little bit of syntactic sugar that would have been appropriate for my use cases.

**Reply by fat-rooster** [reply]
@HaloFour Oh, absolutely. source-only attributes is a much bigger thing, but that also makes it a much wider thing. I am under no illusions that this is top priority, or even particularly important. I just don't see any down sides, and it would make my code slightly cleaner.
It is a new syntax, and keyword, but I cannot fathom what else could possibly go between an interface and the comma after it instead, so that shouldn't be as big a deal as it sounds.

**Reply by parallelbgls** [reply]
In my open source project, there is an implement that copy paste much codes in two seperated classes.

```C#
public abstract class EventHandlerConnector<TParamIn, TParamOut> : ChannelHandlerAdapter, IConnectorWithController<TParamIn, TParamOut> where TParamIn : class

public abstract class BaseConnector<TParamIn, TParamOut> : IConnectorWithController<TParamIn, TParamOut> where TParamIn : class
```

Both implementation of IConnectorWithController in EventHandlerConnector and BaseConnector are the same.

Unfortunatlly EventHandlerConnector use dotnetty base class so I have to copy paste implementation of IConnectorWithController between EventHandlerConnctor and BaseConnector.

Is it possible to auto implement IConnectorWithController using an abstract class, like:

```C#
public abstract class EventHandlerConnector<TParamIn, TParamOut> : ChannelHandlerAdapter, IConnectorWithController<TParamIn, TParamOut> implementby BaseConnectorWithController<TParamIn, TParamOut> where TParamIn : class

public abstract class BaseConnector<TParamIn, TParamOut> : IConnectorWithController<TParamIn, TParamOut> implementby BaseConnectorWithController<TParamIn, TParamOut> where TParamIn : class

public abstract class BaseConnectorWithController<TParamIn, TParamOut>: IConnectorWithController<TParamIn, TParamOut> where TParamIn : class
```

Source code:

[BaseConnector.cs](https://github.com/parallelbgls/Modbus.Net/blob/master/Modbus.Net/Modbus.Net/Connector/BaseConnector.cs)

[EventHandlerConnector.cs](https://github.com/parallelbgls/Modbus.Net/blob/master/Modbus.Net/Modbus.Net/Connector/EventHandlerConnector.cs)
